### PR TITLE
Add Finitary funsor for representing lazy op application

### PR DESCRIPTION
**Pull request by eb8680**
Addresses #351. ~~Blocked by #421, #422.~~

This PR adds a new term `funsor.terms.Finitary` that represents lazy op application, and refactors `funsor.terms.Einsum` into a new op `EinsumOp` that makes use of the `find_domain` extensibility added in #422.

Tested:
- Exercised a bit by existing `Einsum` tests

### CODE REVIEW

**Review by fritzo — APPROVED** [review]
LGTM

File: funsor/tensor.py [hunk 1]
```diff
@@ -695,6 +696,18 @@ def eager_getitem_tensor_tensor(op, lhs, rhs):
     return Tensor(data, inputs, lhs.dtype)
 
 
+# TODO handle variable length Tuple
+@eager.register(Finitary, Op, typing.Tuple[Tensor])
```
Comment (eb8680) [inline funsor/tensor.py:700]: Note the use of `typing.Tuple` in head patterns like this is not currently supported, and will have to be worked around like we did with `Contraction`.  This is what got me thinking about #355, #433 and #434 again.

Comment (eb8680) [reply funsor/tensor.py]: Added support for this in #451.

File: test/test_distribution.py [hunk 2]
```diff
@@ -639,7 +639,7 @@ def test_mvn_affine_einsum():
     data = dict(x=Tensor(randn(2, 2)), y=Tensor(randn(())))
     with lazy:
         d = to_funsor(random_mvn((), 3), Real)
-        d = d(value=Einsum("abc,bc->a", c, x) + y)
+        d = d(value=Einsum("abc,bc->a", (c, x)) + y)
```
Comment (fritzo) [inline test/test_distribution.py:642]: nit: if `Einsum` is just a wrapper, how about letting it match the standard `np.einsum` syntax via `*args`?

File: funsor/tensor.py [hunk 3]
```diff
@@ -1019,7 +1027,30 @@ def max_and_argmax(x: Reals[8]) -> Tuple[Real, Bint[8]]:
...
+        for name, size in zip(ein_input, x.shape):
+            other_size = size_dict.setdefault(name, size)
+            if other_size != size:
+                raise ValueError(
+                    "Size mismatch at {}: {} vs {}".format(name, size, other_size)
+                )
+    return Reals[tuple(size_dict[d] for d in ein_output)]
+
+
+def Einsum(equation, operands):
```
Comment (fritzo) [inline funsor/tensor.py:1053]: nit: How about standardizing as `def Einsum(equaiton, *operands):`?

**Review by fritzo — APPROVED** [review]
This is a great simplification!

Only one nit on keeping the standard `np.einsum` interface

**Pull request merged**